### PR TITLE
Reader: combine consecutive x-posts

### LINKED ISSUE #14722: Reader: fix x-posts filtering to get following state from Redux
We previously filtered x-posts in the Reader based on what blogs you were already following (to prevent seeing the same x-post multiple times in succession).

This functionality is now broken after the move away from Flux. After Redux-ifying streams, we should revisit this.

### PR DESCRIPTION
**Pull request by bluefuton**
Combine consecutive x-posts to the same post or comment.

Part of #14722.

27597241-07d921ae-5b1f-11e7-8ca3-38cf57c97a65

### To test

Visit a stream with x-posts (like the team stream). Make sure that consecutive x-posts for the same original post have been combined like this:

screen shot 2018-07-04 at 15 08 58

### CODE REVIEW

**Comment by bluefuton** [comment]
Todo:
- [ ] separate reducer for x-post URLs
- [ ] tests
- [ ] cleanup combined card style implementation

File: client/state/data-layer/wpcom/read/streams/index.js [hunk 1]
```diff
@@ -228,6 +229,8 @@ export function handlePage( action, data ) {
 		...keyForPost( post ),
 		date: post[ dateProperty ],
 		...( post.comments && { comments: map( post.comments, 'ID' ).reverse() } ), // include comments for conversations
+		url: post.URL,
+		xPostMetadata: isXPost( post ) ? XPostHelper.getXPostMetadata( post ) : null,
```
Comment (blowery) [inline client/state/data-layer/wpcom/read/streams/index.js]: might just push the `isXPost` check down into getXPostMetadata (assuming it returns null if there's no metadata)

Comment (bluefuton) [reply client/state/data-layer/wpcom/read/streams/index.js]: Thanks for the suggestion! Moved in f18fc12.

**Comment by westi** [comment]
This looks great, it will be even better when we can follow up and make the url of the combined card point to the source of the X-Posts/X-Comments but we can leave that for a future change.

File: client/reader/stream/utils.js [hunk 2]
```diff
@@ -38,6 +38,19 @@ export function sameDay( postKey1, postKey2 ) {
 	return moment( postKey1.date ).isSame( postKey2.date, 'day' );
 }
 
+export function sameXPost( postKey1, postKey2 ) {
```
Comment (jsnmoon) [inline client/reader/stream/utils.js]: `someXPost` -> `isSameXPost`?

File: client/state/reader/streams/utils.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * Internal dependencies
+ */
+import { sameXPost } from 'reader/stream/utils';
+
+/**
+ * Add the post URL from a duplicate x-post to an existing post key
+ *
+ * @param {object} postKey1 First post key
+ * @param {object} postKey2 Second (duplicate) post key
+ * @return {array} Array of post key objects
```
Comment (jsnmoon) [inline client/state/reader/streams/utils.js]: Doesn't this return an Object instead of an Array?

**Comment by jsnmoon** [comment]
Aside: Can any wpcom blog create x-posts? Or is the functionality reserved for P2's?

File: client/reader/stream/utils.js (hunk 2 above)
Comment (bluefuton) [reply client/reader/stream/utils.js]: I prefer that too! It's preceded by two functions called `sameSite` and `sameDate` so I've just gone with `sameXPost` for consistency's sake.

**Comment by bluefuton** [comment]
> Aside: Can any wpcom blog create x-posts? Or is the functionality reserved for P2's?

I believe it's only available on multisite installs with the O2/P2 theme:

https://github.com/Automattic/o2/blob/master/inc/xposts.php

File: client/state/reader/streams/utils.js (hunk 3 above)
Comment (bluefuton) [reply client/state/reader/streams/utils.js]: You're right! It returns a post key object. Good spot - I'll fix.

**Pull request merged**